File: sccp/chains/ton/scripts/encode_attester_quorum_proof.mjs
// Encode SCCP AttesterQuorum proof bytes for SORA `pallet-sccp`.
//
// Proof format:
// - version: u8 = 1
// - signatures: SCALE(Vec<[u8;65]>) where each signature is r||s||v (v may be 0/1 or 27/28)
//
// Signatures are over:
//   attestHash = keccak256("sccp:attest:v1" || messageId)
// (raw digest signature, no EIP-191 prefix)
//
// Usage:
//   node scripts/encode_attester_quorum_proof.mjs --message-id 0x.. --sig 0x.. --sig 0x..
//   node scripts/encode_attester_quorum_proof.mjs --message-id 0x.. --privkey 0x.. --privkey 0x..
//
import { Wallet, keccak256, concat, getBytes, toUtf8Bytes } from "ethers";
import { Buffer } from "node:buffer";

function usageAndExit(code) {
  const msg = [
    "Usage:",
    "  node scripts/encode_attester_quorum_proof.mjs --message-id 0x<32-byte> --sig 0x<65-byte> [--sig ...]",
    "  node scripts/encode_attester_quorum_proof.mjs --message-id 0x<32-byte> --privkey 0x<32-byte> [--privkey ...]",
  ].join("\n");
  // eslint-disable-next-line no-console
  console.error(msg);
  process.exit(code);
}

function parseArgs(argv) {
  let messageId = null;
  const sigs = [];
  const privkeys = [];
  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    if (a === "--message-id") {
      const next = argv[i + 1];
      if (next === undefined || next.startsWith("--")) usageAndExit(2);
      messageId = next;
      i += 1;
      continue;
    }
    if (a === "--sig") {
      const next = argv[i + 1];
      if (next === undefined || next.startsWith("--")) usageAndExit(2);
      sigs.push(next);
      i += 1;
      continue;
    }
    if (a === "--privkey") {
      const next = argv[i + 1];
      if (next === undefined || next.startsWith("--")) usageAndExit(2);
      privkeys.push(next);
      i += 1;
      continue;
    }
    usageAndExit(2);
  }
  return { messageId, sigs, privkeys };
}

function normalizeHex(hex, label) {
  if (typeof hex !== "string" || !hex.startsWith("0x")) {
    throw new Error(`${label} must be 0x-prefixed hex`);
  }
  const raw = hex.slice(2);
  if (raw.length % 2 !== 0) {
    throw new Error(`${label} must have an even number of hex digits`);
  }
  if (!/^[0-9a-fA-F]*$/.test(raw)) {
    throw new Error(`${label} must contain only hex digits`);
  }
  return raw.toLowerCase();
}

function assertHexBytes(hex, expectedLen, label) {
  const raw = normalizeHex(hex, label);
  const b = getBytes(`0x${raw}`);
  if (b.length !== expectedLen) {
    throw new Error(`${label} must be ${expectedLen} bytes, got ${b.length}`);
  }
  return b;
}

// SCALE compact-encode a u32 (supports values < 2^30).
function scaleCompactU32(n) {
  if (!Number.isInteger(n) || n < 0) throw new Error("length must be a non-negative integer");
  if (n < 1 << 6) {
    return Uint8Array.from([(n << 2) | 0]);
  }
  if (n < 1 << 14) {
    const v = (n << 2) | 1;
    return Uint8Array.from([v & 0xff, (v >> 8) & 0xff]);
  }
  if (n < 1 << 30) {
    const v = (n << 2) | 2;
    return Uint8Array.from([v & 0xff, (v >> 8) & 0xff, (v >> 16) & 0xff, (v >> 24) & 0xff]);
  }
  throw new Error("length too large for compact u32 encoding");
}

function hexFromBytes(bytes) {
  return "0x" + Buffer.from(bytes).toString("hex");
}

function main() {
  const { messageId, sigs, privkeys } = parseArgs(process.argv);
  if (!messageId) usageAndExit(2);
  if (sigs.length > 0 && privkeys.length > 0) {
    throw new Error("use either --sig or --privkey, not both");
  }
  if (sigs.length === 0 && privkeys.length === 0) {
    throw new Error("at least one --sig or --privkey is required");
  }

  const messageIdBytes = assertHexBytes(messageId, 32, "messageId");
  const attestHash = keccak256(concat([toUtf8Bytes("sccp:attest:v1"), messageIdBytes]));

  const outSigs = [];
  const signers = [];
  if (privkeys.length > 0) {
    for (const k of privkeys) {
      const pk = typeof k === "string" && k.startsWith("0x") ? k : `0x${k}`;
      const normalizedPk = hexFromBytes(assertHexBytes(pk, 32, "private key"));
      const w = new Wallet(normalizedPk);
      const sig = w.signingKey.sign(attestHash).serialized; // 65-byte r||s||v (v=27/28)
      outSigs.push(assertHexBytes(sig, 65, "signature"));
      signers.push(w.address);
    }
  } else {
    for (const s of sigs) {
      outSigs.push(assertHexBytes(s, 65, "signature"));
    }
  }

  const len = scaleCompactU32(outSigs.length);
  const proofBytes = concat([Uint8Array.from([1]), len, ...outSigs]);
  const proofHex = hexFromBytes(proofBytes);
  const proofBase64 = Buffer.from(proofBytes).toString("base64");

  const out = {
    message_id: messageId,
    attest_hash: attestHash,
    signers: signers.length > 0 ? signers : undefined,
    proof_hex: proofHex,
    proof_base64: proofBase64,
  };

  // eslint-disable-next-line no-console
  console.log(JSON.stringify(out, null, 2));
}

try {
  main();
} catch (e) {
  // eslint-disable-next-line no-console
  console.error(`error: ${e.message}`);
  process.exit(1);
}

File: sccp/chains/sol/src/bin/encode_attester_quorum_proof.rs
use sccp_sol::{attest_hash, H256};
use std::env;

fn decode_hex_nibble(b: u8) -> Option<u8> {
    match b {
        b'0'..=b'9' => Some(b - b'0'),
        b'a'..=b'f' => Some(b - b'a' + 10),
        b'A'..=b'F' => Some(b - b'A' + 10),
        _ => None,
    }
}

fn decode_hex_0x(s: &str) -> Result<Vec<u8>, String> {
    let s = s.strip_prefix("0x").unwrap_or(s);
    if s.len() % 2 != 0 {
        return Err("hex string must have even length".to_owned());
    }
    let bytes = s.as_bytes();
    let mut out = Vec::with_capacity(s.len() / 2);
    for i in (0..bytes.len()).step_by(2) {
        let hi = decode_hex_nibble(bytes[i]).ok_or("invalid hex")?;
        let lo = decode_hex_nibble(bytes[i + 1]).ok_or("invalid hex")?;
        out.push((hi << 4) | lo);
    }
    Ok(out)
}

fn encode_compact_u32(n: u32) -> Result<Vec<u8>, String> {
    // SCALE compact encoding for values < 2^30.
    if n < (1 << 6) {
        return Ok(vec![((n << 2) as u8) | 0]);
    }
    if n < (1 << 14) {
        let v = (n << 2) | 1;
        return Ok(vec![(v & 0xff) as u8, ((v >> 8) & 0xff) as u8]);
    }
    if n < (1 << 30) {
        let v = (n << 2) | 2;
        return Ok(vec![
            (v & 0xff) as u8,
            ((v >> 8) & 0xff) as u8,
            ((v >> 16) & 0xff) as u8,
            ((v >> 24) & 0xff) as u8,
        ]);
    }
    Err("value too large for compact u32 encoding".to_owned())
}

fn hex_encode(bytes: &[u8]) -> String {
    const HEX: &[u8; 16] = b"0123456789abcdef";
    let mut out = String::with_capacity(2 + bytes.len() * 2);
    out.push_str("0x");
    for &b in bytes {
        out.push(HEX[(b >> 4) as usize] as char);
        out.push(HEX[(b & 0x0f) as usize] as char);
    }
    out
}

fn usage() -> ! {
    eprintln!(
        "Usage:\n  cargo run --bin encode_attester_quorum_proof -- --message-id 0x<32b> --sig 0x<65b> [--sig ...]"
    );
    std::process::exit(2);
}

fn main() {
    let mut message_id: Option<String> = None;
    let mut sigs: Vec<String> = Vec::new();

    let mut it = env::args().skip(1);
    while let Some(a) = it.next() {
        match a.as_str() {
            "--message-id" => message_id = it.next(),
            "--sig" => sigs.push(it.next().unwrap_or_default()),
            "--help" | "-h" => usage(),
            _ => usage(),
        }
    }

    let message_id = message_id.unwrap_or_else(|| usage());
    if sigs.is_empty() {
        usage();
    }

    let msg_bytes = decode_hex_0x(&message_id).unwrap_or_else(|e| {
        eprintln!("invalid message id: {e}");
        std::process::exit(2);
    });
    if msg_bytes.len() != 32 {
        eprintln!("message id must be 32 bytes");
        std::process::exit(2);
    }
    let mut mid: H256 = [0u8; 32];
    mid.copy_from_slice(&msg_bytes);

    let ah = attest_hash(&mid);

    let mut proof: Vec<u8> = Vec::new();
    proof.push(1u8); // version
    proof.extend_from_slice(&encode_compact_u32(sigs.len() as u32).unwrap_or_else(|e| {
        eprintln!("encode compact failed: {e}");
        std::process::exit(2);
    }));
    for s in sigs.iter() {
        let sb = decode_hex_0x(s).unwrap_or_else(|e| {
            eprintln!("invalid signature: {e}");
            std::process::exit(2);
        });
        if sb.len() != 65 {
            eprintln!("signature must be 65 bytes");
            std::process::exit(2);
        }
        proof.extend_from_slice(&sb);
    }

    println!("{{");
    println!("  \"message_id\": \"{}\",", hex_encode(&mid));
    println!("  \"attest_hash\": \"{}\",", hex_encode(&ah));
    println!("  \"proof_hex\": \"{}\"", hex_encode(&proof));
    println!("}}");
}
